File: ruby/vendor1.rb


module Vendor1

  def Vendor1.createDeviceA(index)
    Vendor1DeviceA.new(index)
  end

  def Vendor1.createDeviceB(index)
    Vendor1DeviceB.new(index)
  end


  class Vendor1Device
    def initialize(index)
      @x = index
    end
  end

  class Vendor1DeviceA < Vendor1Device
    def op1(param)
      @x += param
    end

    def op2(param)
      @x
    end
  end

  class Vendor1DeviceB < Vendor1Device
    def op1(param)
      @x *= param
    end

    def op2(x, y)
      @x * x + y
    end
  end

end

File: javascript/vendor1.js

class Vendor1Device {
  constructor(index) { this.x = index }
}

class Vendor1DeviceA extends Vendor1Device {
  op1(param) { this.x += param }
  op2(param) { return this.x }
}

class Vendor1DeviceB extends Vendor1Device {
  op1(param) { this.x *= param }
  op2(x, y) { return this.x * x + y }
}

class Vendor1 {
  createDeviceA(index) { return new Vendor1DeviceA(index) }
  createDeviceB(index) { return new Vendor1DeviceB(index) }
}


module.exports = Vendor1
